### ISSUE TITLE
Start unpriviledged container without root with an apparmor profile

### ISSUE THREAD

**Issue by raldone01** [lxc/lxc#4358]
# Required information

 * Distribution: arch linux
 * Distribution version: rolling

```
❯ lxc-start --version
5.0.3
❯ uname -a
Linux argon 6.5.6-arch2-1 #1 SMP PREEMPT_DYNAMIC Sat, 07 Oct 2023 08:14:55 +0000 x86_64 GNU/Linux
❯ cat /proc/self/cgroup
0::/user.slice/user-1000.slice/session-3.scope
```

# Issue description

```
# whoami
root
# lxc-create -n lxc-arch -t download -- --dist archlinux --release current --arch amd64
# lxc-start -s lxc.apparmor.profile=generated -n lxc-arch -F
Working and starts fine.
❯ su unpriviledged_user
❯ lxc-create -n lxc-arch -t download -- --dist archlinux --release current --arch amd64
❯ lxc-start -s lxc.apparmor.profile=generated -n lxc-arch -F
lxc-start: lxc-arch: ../src/lxc/lsm/apparmor.c: make_apparmor_namespace: 869 Permission denied - Error creating AppArmor namespace: /sys/kernel/security/apparmor/policy/namespaces/lxc-lxc-arch_<-home-main-.local-share-lxc>
lxc-start: lxc-arch: ../src/lxc/lsm/apparmor.c: apparmor_prepare: 1088 Failed to load generated AppArmor profile
lxc-start: lxc-arch: ../src/lxc/start.c: lxc_init: 876 Failed to initialize LSM
lxc-start: lxc-arch: ../src/lxc/start.c: __lxc_start: 2027 Failed to initialize container "lxc-arch"
lxc-start: lxc-arch: ../src/lxc/tools/lxc_start.c: main: 306 The container failed to start
lxc-start: lxc-arch: ../src/lxc/tools/lxc_start.c: main: 311 Additional information can be obtained by setting the --logfile and --logpriority options
❯ lxc-start -s lxc.apparmor.profile=unconfined -n lxc-arch -F
Working and starts fine.
```

**Comment by mihalicyn** [comment]
Hi @raldone01 

Sorry for long delay with reply.

>Running an unprivileged user with app armor confinement fails.
>Is it possible to run an unprivileged container with confinement?

You need to have `CAP_MAC_ADMIN` in the current user namespace to manage policies (https://github.com/torvalds/linux/blob/ffd2cb6b718e189e7e2d5d0c19c25611f92e061a/security/apparmor/policy.c#L839)

So, no, you can't manage apparmor from an unprivileged user. As far as I remember there is a feature (apparmor stacking) that allows you to create a stacked profiles inside the user namespace. But it is another story anyways.

If you are unprivileged, there is not so much sense to user AppArmor at all.

Why do you want to do that?

**Comment by mihalicyn** [comment]
> So using an unprivileged user with USER_NS was less safe than using a privileged user with apparmor confinement.

I don't think so. unprivileged user + user namespace is a very-very safe thing. And this is an upstream Linux kernel default option, so all the vulnerability researchers are paying attention to this configuration all the time. That's why this combo has so big history of CVEs. But that's a sign of a big attention to it! It does not mean that this is an ugly combo from a security standpoint.

If we speak about AppArmor, I makes sense to say that AppArmor is just a specific LSM module for Linux kernel. Not so many eyes looking at it and not so many researchers as a consequence.

It's a classical https://en.wikipedia.org/wiki/Survivorship_bias ;-)

>Which is safer?
>
>    A: Container with privileged user and confinement.

You mean "privileged container" right? Privileged container means that container is not using user namespace at all.
It means that root inside the container IS root on the host. But it's confined with AppArmor.

>    B: Container with unprivileged user (USER_NS) and unconfined.

Yep, it means that root inside the container, is just a regular user on the host. All the security tricks in this case comes
from the vanilla upstream Linux kernel, without any custom things like AppArmor. (Very popular and well tested!)

>    C: Container with privileged user (USER_NS) and confinement. (Is that even possible?)

hm, yep, there is a confusion about naming. Let's try to sort this out:
- privileged container means that container does NOT use user namespaces
- privileged user (usually) means that user has a capabilities on the host (in the initial user namespace)

Once you create a user namespace, then you lose all the capabilities in the initial user namespace. For example:
```
# create user namespace and enter to it
root@host $ unshare -Ur
```
or
```
# create user namespace and enter to it
user@host $ unshare -Ur
```

There is no difference from the capabilities perspective at all! The only difference would be that in first case uid 0 inside the user namespace will be uid 0 outside. In 2nd case uid 0 inside the user namespace will be uid 1000 (if a user "user" has uid 1000). But from the permissions perspective there is no difference at all! Because uid 0  does not mean that you have a power of root. Power of root is about having capabilities, not UID!

Let's get back to your list of options A, B, C and ordering. I can't say anything about C because it's not clear what it means.
But I definitely say that B is more secure than A. Using unconfined user namespace is safer.

I'm pretty sure that Stéphane has something to add :-)
cc @stgraber

**Comment by hallyn** [comment]
Thanks for the clarification - fwiw, [a] is imo not safe.  [c] certainly is possible.  There is also [D]: Container launched by un unpriv user and confined.  That is, of course, the safest.

**Comment by raldone01** [comment]
So to achieve D I would have to give the unprivileged user the `CAP_MAC_ADMIN` privilege right? Otherwise I run into my initial issue.
Do you happen to know how I can give my user the `CAP_MAC_ADMIN` or is it a bad idea to this to my regular user account?

**Comment by hallyn** [comment]
Sorry I probably shouldn't have jumped in.

Yes, to load newly generated or ad-hoc policies requires CAP_MAC_ADMIN, and you do not want to give that to users.

What you can do is load custom policies for them in advance.  Then the users can start containers using those policies at any time.  This is generally the norm: unprivileged containers are not actually un-confined, but are confined by a rather general policy.

**Comment by hallyn** [comment]
```
ubuntu@puzzlefs:~$ lxc-create -t oci -n puzzle1 -- --url oci:/home/ubuntu/ptest/oci:alpine-puzzle
[...]
Copying config 44136fa355 [--------------------------------------] 0.0b / 2.0b | 0.0 b/s
FATA[0001] reading config blob sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a: open /home/ubuntu/ptest/oci/blobs/sha256/44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a: no such file or directory
lxc-create: puzzle1: ../src/lxc/lxccontainer.c: create_run_template: 1628 Failed to create container from template
lxc-create: puzzle1: ../src/lxc/tools/lxc_create.c: main: 317 Failed to create container puzzle1
```

**Comment by raldone01** [comment]
@hallyn I don't quite understand your comment...

**Comment by hallyn** [comment]
sorry i don't think that was intended for this issue!

**Comment by mihalicyn** [comment]
From what I see, we have clarified everything here. Closing for now.

Feel free to reopen if you have any questions.
